### PR TITLE
HDDS-1347. In OM HA getS3Secret call Should happen only leader OM.

### PR DESCRIPTION
**Pull request by bharatviswa504**

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 24 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 10 | Maven dependency ordering for branch |
| +1 | mvninstall | 991 | trunk passed |
| +1 | compile | 96 | trunk passed |
| +1 | checkstyle | 26 | trunk passed |
| +1 | mvnsite | 63 | trunk passed |
| +1 | shadedclient | 714 | branch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 101 | trunk passed |
| +1 | javadoc | 57 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 13 | Maven dependency ordering for patch |
| +1 | mvninstall | 65 | the patch passed |
| +1 | compile | 92 | the patch passed |
| +1 | cc | 92 | the patch passed |
| +1 | javac | 92 | the patch passed |
| -0 | checkstyle | 20 | hadoop-ozone: The patch generated 3 new + 0 unchanged - 0 fixed = 3 total (was 0) |
| +1 | mvnsite | 55 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 729 | patch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 120 | the patch passed |
| +1 | javadoc | 51 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 36 | common in the patch passed. |
| -1 | unit | 41 | ozone-manager in the patch failed. |
| +1 | asflicense | 25 | The patch does not generate ASF License warnings. |
| | | 3385 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.om.ratis.TestOzoneManagerRatisServer |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-670/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/670 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  findbugs  checkstyle  cc  |
| uname | Linux bc6beb534300 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / ec82e4c |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| findbugs | v3.1.0-RC1 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-670/1/artifact/out/diff-checkstyle-hadoop-ozone.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-670/1/artifact/out/patch-unit-hadoop-ozone_ozone-manager.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-670/1/testReport/ |
| Max. process+thread count | 438 (vs. ulimit of 5500) |
| modules | C: hadoop-ozone/common hadoop-ozone/ozone-manager U: hadoop-ozone |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-670/1/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 40 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 68 | Maven dependency ordering for branch |
| +1 | mvninstall | 1174 | trunk passed |
| +1 | compile | 110 | trunk passed |
| +1 | checkstyle | 31 | trunk passed |
| +1 | mvnsite | 69 | trunk passed |
| +1 | shadedclient | 785 | branch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 114 | trunk passed |
| +1 | javadoc | 64 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 12 | Maven dependency ordering for patch |
| +1 | mvninstall | 76 | the patch passed |
| +1 | compile | 96 | the patch passed |
| +1 | cc | 96 | the patch passed |
| +1 | javac | 96 | the patch passed |
| -0 | checkstyle | 23 | hadoop-ozone: The patch generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0) |
| +1 | mvnsite | 58 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 762 | patch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 116 | the patch passed |
| +1 | javadoc | 55 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 35 | common in the patch passed. |
| +1 | unit | 52 | ozone-manager in the patch passed. |
| +1 | asflicense | 28 | The patch does not generate ASF License warnings. |
| | | 3832 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-670/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/670 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  findbugs  checkstyle  cc  |
| uname | Linux 9da069f35dd1 3.13.0-153-generic #203-Ubuntu SMP Thu Jun 14 08:52:28 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 04f1db8 |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| findbugs | v3.1.0-RC1 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-670/2/artifact/out/diff-checkstyle-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-670/2/testReport/ |
| Max. process+thread count | 392 (vs. ulimit of 5500) |
| modules | C: hadoop-ozone/common hadoop-ozone/ozone-manager U: hadoop-ozone |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-670/2/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by bharatviswa504** [comment]
Rebased this. It is ready for review.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java [hunk 1]
```diff
@@ -2683,6 +2683,14 @@ public void deleteS3Bucket(String s3BucketName) throws IOException {
    * {@inheritDoc}
    */
   public S3SecretValue getS3Secret(String kerberosID) throws IOException{
+    UserGroupInformation user = ProtobufRpcEngine.Server.getRemoteUser();
+
+    // Check whether user name passed is matching with the current user or not.
```
Comment (arp7) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java]: Why is this new check required?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/s3/security/S3GetSecretRequest.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+          .newBuilder().setS3Secret(S3Secret.newBuilder()
+          .setAwsSecret(awsSecret).setKerberosID(kerberosID));
+
+      if (s3SecretValue == null) {
+        omClientResponse =
+            new S3GetSecretResponse(new S3SecretValue(kerberosID, awsSecret),
+            omResponse.setGetS3SecretResponse(getS3SecretResponse).build());
+      } else {
+        // As when it already exists, we don't need to add to DB again. So
+        // set the value to null.
```
Comment (arp7) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/s3/security/S3GetSecretRequest.java]: Can you use Optional in place of null in the response object. In general usage of null in code is a potential NPE. :)

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java (hunk 1 above)
Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java]: This was based on some related review comments when original change for getS3Secret is implemented.
So, I implemented it as part of this.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/s3/security/S3GetSecretRequest.java (hunk 2 above)
Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/s3/security/S3GetSecretRequest.java]: For this reason, I have declared this as @Nullable.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 38 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 64 | Maven dependency ordering for branch |
| +1 | mvninstall | 614 | trunk passed |
| +1 | compile | 349 | trunk passed |
| +1 | checkstyle | 61 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 797 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 148 | trunk passed |
| 0 | spotbugs | 419 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 612 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 27 | Maven dependency ordering for patch |
| +1 | mvninstall | 536 | the patch passed |
| +1 | compile | 357 | the patch passed |
| +1 | cc | 357 | the patch passed |
| +1 | javac | 357 | the patch passed |
| +1 | checkstyle | 64 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 632 | patch has no errors when building and testing our client artifacts. |
| -1 | javadoc | 96 | hadoop-ozone generated 1 new + 26 unchanged - 0 fixed = 27 total (was 26) |
| +1 | findbugs | 665 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 329 | hadoop-hdds in the patch passed. |
| -1 | unit | 3436 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 44 | The patch does not generate ASF License warnings. |
| | | 9101 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.scm.pipeline.TestPipelineManagerMXBean |
|   | hadoop.ozone.client.rpc.TestOzoneAtRestEncryption |
|   | hadoop.ozone.ozShell.TestOzoneShell |
|   | hadoop.ozone.scm.node.TestQueryNode |
|   | hadoop.ozone.container.TestContainerReplication |
|   | hadoop.ozone.om.TestOzoneManagerHA |
|   | hadoop.ozone.client.rpc.Test2WayCommitInRatis |
|   | hadoop.ozone.dn.scrubber.TestDataScrubber |
|   | hadoop.ozone.container.server.TestSecureContainerServer |
|   | hadoop.ozone.ozShell.TestS3Shell |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-670/6/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/670 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle cc |
| uname | Linux 1cbd7ceb0ff2 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 8fc6567 |
| Default Java | 1.8.0_222 |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-670/6/artifact/out/diff-javadoc-javadoc-hadoop-ozone.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-670/6/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-670/6/testReport/ |
| Max. process+thread count | 4302 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/common hadoop-ozone/common hadoop-ozone/integration-test hadoop-ozone/ozone-manager U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-670/6/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 76 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 23 | Maven dependency ordering for branch |
| +1 | mvninstall | 663 | trunk passed |
| +1 | compile | 392 | trunk passed |
| +1 | checkstyle | 78 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 961 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 183 | trunk passed |
| 0 | spotbugs | 450 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 663 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 33 | Maven dependency ordering for patch |
| +1 | mvninstall | 618 | the patch passed |
| +1 | compile | 416 | the patch passed |
| +1 | cc | 416 | the patch passed |
| +1 | javac | 416 | the patch passed |
| +1 | checkstyle | 85 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 779 | patch has no errors when building and testing our client artifacts. |
| -1 | javadoc | 96 | hadoop-ozone generated 1 new + 26 unchanged - 0 fixed = 27 total (was 26) |
| +1 | findbugs | 691 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 352 | hadoop-hdds in the patch passed. |
| -1 | unit | 2280 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 42 | The patch does not generate ASF License warnings. |
| | | 8660 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.client.rpc.TestCommitWatcher |
|   | hadoop.ozone.container.ozoneimpl.TestOzoneContainer |
|   | hadoop.ozone.client.rpc.TestOzoneClientRetriesOnException |
|   | hadoop.ozone.client.rpc.TestBlockOutputStreamWithFailures |
|   | hadoop.ozone.container.server.TestSecureContainerServer |
|   | hadoop.hdds.scm.pipeline.TestRatisPipelineCreateAndDestory |
|   | hadoop.ozone.om.TestOzoneManagerHA |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-670/7/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/670 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle cc |
| uname | Linux 7f29569600a6 4.15.0-54-generic #58-Ubuntu SMP Mon Jun 24 10:55:24 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 69ddb36 |
| Default Java | 1.8.0_222 |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-670/7/artifact/out/diff-javadoc-javadoc-hadoop-ozone.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-670/7/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-670/7/testReport/ |
| Max. process+thread count | 4912 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/common hadoop-ozone/common hadoop-ozone/integration-test hadoop-ozone/ozone-manager U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-670/7/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by bharatviswa504** [comment]
Thank You @arp7 for the review.
I have committed this to the trunk.

**Pull request merged**